File: Cantera/src/oneD/refine.cpp

#include <map>
#include <algorithm>
#include "Resid1D.h"

#include "refine.h"

using namespace std;

namespace Cantera {


    template<class M>
    static bool has_key(const M& m, int j) {
        if (m.find(j) != m.end()) return true;
        return false;
    }


    /**
     * Return the square root of machine precision.
     */
    static doublereal eps() {
        doublereal e = 1.0;
        while (1.0 + e != 1.0) e *= 0.5;
        return sqrt(e);
    }

    
    Refiner::Refiner(Resid1D& domain) :
        m_ratio(10.0), m_slope(0.8), m_curve(0.8), m_min_range(0.01),
        m_domain(&domain)
    {
        m_nv = m_domain->nComponents();
        m_active.resize(m_nv, true);
        m_thresh = eps();
    }
            

    int Refiner::analyze(int n, const doublereal* z, 
        const doublereal* x) {

        if (m_domain->nPoints() <= 1) return 0;
        m_nv = m_domain->nComponents();

        //m_ok = false;

        // check consistency
        if (n != m_domain->nPoints()) return -1;

        m_loc.clear();
        m_c.clear();

        /**
         * find locations where cell size ratio is too large.
         */
        int j;
        vector_fp dz(n-1, 0.0);
        dz[0] = z[1] - z[0];
        for (j = 1; j < n-1; j++) {
            dz[j] = z[j+1] - z[j];
            if (dz[j] > m_ratio*dz[j-1]) {
                m_loc[j] = 1;
                m_c["point "+int2str(j)] = 1;
            }
            if (dz[j] < dz[j-1]/m_ratio) {
                m_loc[j-1] = 1;
                m_c["point "+int2str(j-1)] = 1;                
            }
        }

        string name;
        doublereal vmin, vmax, smin, smax, aa, ss;
        doublereal dmax, r;
        vector_fp v(n), s(n-1);
        for (int i = 0; i < m_nv; i++) {
            //cout << i << "   " << m_nv << "  " << m_active[i] << endl;
            if (m_active[i]) {
                name = m_domain->componentName(i);
            
                // get component i at all points
                for (j = 0; j < n; j++) v[j] = value(x, i, j); 

                // slope of component i
                for (j = 0; j < n-1; j++)
                    s[j] = (value(x, i, j+1) - value(x, i, j))/
                           (z[j+1] - z[j]);

                // find the range of values and slopes
            
                vmin = *min_element(v.begin(), v.end());
                vmax = *max_element(v.begin(), v.end());
                smin = *min_element(s.begin(), s.end());
                smax = *max_element(s.begin(), s.end());

                // max absolute values of v and s
                aa = fmaxx(abs(vmax), abs(vmin));
                ss = fmaxx(abs(smax), abs(smin));


                // refine based on component i only if the range of v is
                // greater than a fraction 'min_range' of max |v|. This
                // eliminates components that consist of small fluctuations
                // on a constant background.

                if ((vmax - vmin) > m_min_range*aa) {

                    // maximum allowable difference in value between
                    // adjacent points.
                
                    dmax = m_slope*(vmax - vmin) + m_thresh;
                    for (j = 0; j < n-1; j++) {
                        r = abs(v[j+1] - v[j])/dmax;
                        if (r > 1.0) {
                            m_loc[j] = 1;
                            m_c[name] = 1;
                        }
                    }
                }

                        
                // refine based on the slope of component i only if the
                // range of s is greater than a fraction 'min_range' of max
                // |s|. This eliminates components that consist of small
                // fluctuations on a constant slope background.
            
                if ((smax - smin) > m_min_range*ss) {

                    // maximum allowable difference in slope between
                    // adjacent points.
                    dmax = m_curve*(smax - smin);
                    for (j = 0; j < n-2; j++) {
                        r = abs(s[j+1] - s[j]) / (dmax + m_thresh/dz[j]);
                        if (r > 1.0) {
                            m_c[name] = 1;
                            m_loc[j] = 1;
                            m_loc[j+1] = 1;
                        }
                        //cout << "at point " << j << " slope r = "
                        //     << r << " for " << name << endl
                        //     << "    threshold = " << m_thresh << endl;
                    }
                }
                //cout << name << "  " << m_curve << "  " << smax << "  " << smin << "   " << ss << "   " << m_min_range << endl;
            }
        }
        return m_loc.size();
    }

    double Refiner::value(const double* x, int i, int j) {
        return x[m_domain->index(i,j)];
    }

    void Refiner::show() {
        int nnew = m_loc.size();
        if (nnew > 0) {
            writelog("Refining grid.  "
                "New points inserted after grid points ");
            map<int, int>::const_iterator b = m_loc.begin();
            for (; b != m_loc.end(); ++b) {
                writelog(int2str(b->first)+" ");
            }
            writelog("\n");
            writelog("to resolve ");
            map<string, int>::const_iterator bb = m_c.begin();
            for (; bb != m_c.end(); ++bb) {
                writelog(string(bb->first)+" ");
            }
            writelog("\n");
        }
    }


    int Refiner::getNewGrid(int n, const doublereal* z, 
        int nn, doublereal* zn) {
        int j;
        int nnew = m_loc.size();
        if (nnew + n > nn) {
            throw CanteraError("Refine::getNewGrid",
                "array size too small.");
            return -1;
        }

        int jn = 0;
        if (m_loc.size() == 0) {
            copy(z, z + n,  zn);
            return 0;
        }

        for (j = 0; j < n - 1; j++) {
            zn[jn] = z[j];
            jn++;
            if (has_key(m_loc, j)) {
                zn[jn] = 0.5*(z[j] + z[j+1]);
                jn++;
            }
        }
        zn[jn] = z[n-1];
        return 0;
    }

//         int npts = znew.size();
//         newsoln.resize(npts*ncomp);
//         newsoln = Numeric.zeros((npts, ncomp),'d')
//         for i in range(ncomp):
//             for j in range(npts):
//                 newsoln[j,i] = interp.interp(znew[j],grid,solution[:,i])

//         return (Numeric.array(znew), Numeric.array(znew), newsoln, self.ok)
                               
    
                
            
    
}

File: Cantera/python/Cantera/refine.py
"""Grid refinement.

Suppose you have a monotonic NumPy array of grid points 'z', and a
solution array soln[j,n] that contains 3 three solution components
denoted 'a', 'b', and 'c', evaluated at the grid points. To refine the
grid based on components 'a' and 'b' but not 'c', do the following.

>>> from refine import Refiner
>>> r = Refiner([(0, 'a'), (1, 'b')])
>>> new_grid, new_soln = r.refine(grid, soln)

"""


import Numeric
import math
from Cantera import CanteraError
from Cantera import interp

def eps():
    """Return the square root of machine precision."""
    e = 1.0
    while 1.0 + e <> 1.0: e = 0.5*e
    return math.sqrt(e)


def delta(f):
    """Given an array f, return an array of the difference in
    adjacent values."""
    n = len(f)
    d = Numeric.zeros(n-1,'d')
    for j in range(n-1):
        d[j] = f[j+1] - f[j]
    return d


def slope(z, f):
    """Given arrays z and f, return an array of the slopes df/dz in
    each interval."""
    n = len(z)
    s = Numeric.zeros(n-1,'d')
    for j in range(n-1):
        s[j] = ((f[j+1] - f[j])/(z[j+1] - z[j]))
    return Numeric.array(s,'d')


class RefineError(CanteraError):
    def __init__(self, msg):
        self.msg = 'Grid refinement error!\n'+msg


class Refiner:
    """Grid refiner.

    Attributes:

    components -- sequence of (number, name) pairs specifying the
    components of the solution to use for grid refinement. The number
    is used to access the component in the solution array, and the
    name is used only for diagnostic messages.

    max_delta -- Maximum tolerated difference in solution values
    between neighboring grid points, expressed as a fraction between 0
    and 1 of the total range of the component over all grid
    points. Default: 0.8 (minimal refinement).

    max_delta_slope -- Maximum tolerated difference in solution slopes
    between neighboring grid intervals, expressed as a fraction between 0
    and 1 of the total range of the component over all grid
    points. Default: 0.8 (minimal refinement).    
    
    """

    def __init__(self, components = [], delta = (2.0, 0.1, 0.2), names = []):
        self.components = components
        self.delta = delta
        self.names = names
        self.loglevel = 2
        self.eps = eps()
        self.min_range = 0.01
        self.direction = 1
        self.fctr = 1.0
        self.ok = 0


    def prune(self, grid = None, solution = None, threshold = None):

        n0 = len(grid)
        g = grid
        sol = solution
        self.fctr = 1.0
        savedir = self.direction
        
        ll = self.loglevel
        #self.loglevel = 0
        j = 1
        while j < len(g)-1:
            g0 = g
            s0 = sol
            pt = g[j]
            nn = len(g)
            
            # remove point j
            g = Numeric.take(g,range(0,j)+range(j+1,nn))

            # remove row j
            sol = Numeric.take(sol, range(0,j)+range(j+1,nn))
            np = len(g)

            self.direction = 1
            gnew, gn, snew, ok = self.refine(g, sol, threshold)
            if (len(gnew) > np):
                g = g0
                sol = s0
                j += 1
                if ll > 0:
                    print 'cannot remove point at ',pt
            else:
                if ll > 0:
                    print 'removed point at ',pt
        self.loglevel = ll
        self.fctr = 0.2
        self.direction = savedir
        return (g, sol)
    
            
    def refine(self, grid = None, solution = None, threshold = None, prune = 1):

        self.ok = 0
        # grid parameters
        n0 = len(grid)
        dz0 = grid[-1] - grid[0]

        maxpts = self.fctr*n0 + 1
        
        ncomp = Numeric.shape(solution)[1]
        
        if threshold:
            self.threshold = threshold
        else:
            self.threshold = self.eps * Numeric.ones(ncomp, 'd')
            
        if Numeric.shape(solution)[0] <> n0:
            raise RefineError('Number of solution points differs from '+
                              'number of grid points.')

        # if the solution components to examine for refinement have
        # not been specified, use all components.
        nc = Numeric.shape(solution)[1]
        if not self.components: self.components = range(nc)

        c = {}
        p = {}

        dz = delta(grid)
        for j in range(1,n0-1):
            if dz[j] > self.delta[0]*dz[j-1]:
                p[j] = 1
                c['point '+`j`] = 1
            if dz[j] < dz[j-1]/self.delta[0]:
                p[j-1] = 1
                c['point '+`j-1`] = 1                

        for i in self.components:
            try:
                name = self.names[i]
            except:
                name = 'component '+`i`
            
            # get component i at all points, and compute its slope
            v = solution[:,i]
            s = slope(grid, v)

            # compute the change in value and slope
            dv = delta(v)
            ds = delta(s)

            # find the range of values and slopes
            vmin = min(v)
            vmax = max(v)
            smin = min(s)
            smax = max(s)

            # max absolute values of v and s
            aa = max((abs(vmax), abs(vmin)))
            ss = max((abs(smax), abs(smin)))


            # refine based on component i only if the range of v is
            # greater than a fraction 'min_range' of max |v|. This
            # eliminates components that consist of small fluctuations
            # on a constant background.

            if (vmax - vmin) > self.min_range*aa:

                # maximum allowable difference in value between
                # adjacent points.
                
                dmax = self.delta[1]*(vmax - vmin) + self.threshold[i]
                for j in range(len(dv)):
                    r = abs(dv[j])/dmax
                    if r > 1.0:
                        p[j] = 1
                        c[name] = 1

                        
            # refine based on the slope of component i only if the
            # range of s is greater than a fraction 'min_range' of max
            # |s|. This eliminates components that consist of small
            # fluctuations on a constant slope background.
            
            if (smax - smin) > self.min_range*ss:

                # maximum allowable difference in slope between
                # adjacent points.
                dmax = self.delta[2]*(smax - smin)
                
                for j in range(len(ds)):
                    r = abs(ds[j]) / (dmax + self.threshold[i]/dz[j])
                    if r > 1:
                        c[name] = 1
                        p[j] = 1
                        p[j+1] = 1

        if len(p) == 0: self.ok = 1

        znew = []
        nnew = len(p)
        nadded = nnew
                
        if self.loglevel > 0:
            if nnew > 0:
                print '\nRefining grid.'
                print 'New points inserted after grid points',

        for j in range(n0 - 1):
            znew.append(grid[j])
            if p.has_key(j):
                if self.loglevel > 0: print j,
                znew.append(0.5*(grid[j] + grid[j+1]))
        if self.loglevel > 0: print
        znew.append(grid[-1])
        if self.loglevel > 0 and nnew > 0:
            print 'to resolve ',
            ck = c.keys()
            for s in ck:
                if s <> ck[-1]:
                    print s+',',
                else:
                    print s,
            print
        
        npts = len(znew)
        
        newsoln = Numeric.zeros((npts, ncomp),'d')
        for i in range(ncomp):
            for j in range(npts):
                newsoln[j,i] = interp.interp(znew[j],grid,solution[:,i])

        return (Numeric.array(znew), Numeric.array(znew), newsoln, self.ok)



def refine(grid = None, solution = None, components = [], delta = (0.8, 1.0), threshold = None):
    """Refine a grid and interpolate the solution onto the new grid."""
    r = Refiner(components = components, delta = delta)
    return r.refine(grid, solution, threshold)


def prune(grid = None, solution = None, components = [], delta = (0.8, 1.0), threshold = None):
    """Remove unneeded points from a grid and solution array."""
    r = Refiner(components = components, delta = delta)
    return r.prune(grid, solution, threshold)



# test it
if __name__ == '__main__':

    grid = Numeric.array([0.0, 0.2, 0.3, 1.0, 4.0])
    soln = Numeric.array([[100.0, 0.4, -9.0], 
                          [500.0, 0.0, -89.0], 
                          [700.0, 0.9, 99.0],
                          [-99.0, 8.0, 77.0],
                          [567.0, 8.0, 0.0]])
    grid_new, soln_new = refine(grid, soln, 
                                components = [0,2], 
                                delta = (0.5, 0.8))
    
    print 'new grid = ',grid_new
    print 'new solution = ',soln_new
    
                
            
    
